internal/credspropagation/common.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package credspropagation
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hmc "github.com/Mirantis/hmc/api/v1alpha1"
+)
+
+type PropagationCfg struct {
+	Ctx             context.Context
```

Comment:
@zerospiel Ok let me elaborate on your notes and explain my decisions:
1. I passed context, because I use the client to get some objects, so I thought it would be better to have the same client and context as in controller. Or you just saying that I should pass context only as a function argument (weird, but ok)?
2. I used struct specifically to simplify function signature. Using constructor and pass the same values as arguments to it will defy that purpose and abstracting away `PropagationCfg` doesn't make much sense to me in this context (we still need that data passed and not internal states will be in there). 
3. And I can still make the 2 public functions a methods which is pretty much the same thing as it is now.
P.S. Linter passed fine in the CI btw, so probably your linter config wasn't fully merged somehow.

Reply:
`containedctx` linter actually exists for a reason, most of the time it is a bad idea to put context in structure. So if this linter is not enabled it should be enabled in GHA.

I there is a very needful need to put context in structs, there should be a comment with justification about that.

ref: https://go.dev/blog/context-and-structs

Reply:
TLDR: the longer fails in your branch, and the jobs are running against the master branch due to the issues with the CI

Reply:
@zerospiel 

Regarding context: if it's somehow confusing and unidiomatic I'll remove it from struct and will pass it as an argument.

Regarding argument list:
For example
I could've written the `PropagateAzureSecrets` like this

```go
func PropagateAzureSecrets(ctx context.Context, client client.Client, mcl *hmc.ManagedCluster, ks *corev1.Secret, sysNS string) error {}
```
But I don't like to handle that many arguments and it doesn't make sense to split this function, so I came up with this `PropagationCfg` struct to avoid it and I can just declare a struct and pass it as a single argument.

With a constructor I will have to do the following:

```go
func NewPropagateCfg(client client.Client, mcl *hmc.ManagedCluster, ks *corev1.Secret, sysNS string) *PropagateCfg {}
```

Which looks exactly that I wanted to avoid here and IMO makes things bit more complicated. If you insist on that I will ditch the struct completely in favor of just arguments.

Reply:
Correct, either the latter or exactly the same that you've already done. So I mark it resolved, thanks